bitcoin/src/address/mod.rs

```diff
@@ -724,41 +779,24 @@ impl<V: NetworkValidation> fmt::Debug for Address<V> {
     }
 }
 
-/// Extracts the bech32 prefix.
-///
-/// # Returns
-/// The input slice if no prefix is found.
-fn find_bech32_prefix(bech32: &str) -> &str {
-    // Split at the last occurrence of the separator character '1'.
-    match bech32.rfind('1') {
-        None => bech32,
-        Some(sep) => bech32.split_at(sep).0,
-    }
-}
-
 /// Address can be parsed only with `NetworkUnchecked`.
 impl FromStr for Address<NetworkUnchecked> {
     type Err = ParseError;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        // try bech32
-        let bech32_network = match find_bech32_prefix(s) {
-            // note that upper or lowercase is allowed but NOT mixed case
-            "bc" | "BC" => Some(Network::Bitcoin),
-            "tb" | "TB" => Some(Network::Testnet), // this may also be signet
-            "bcrt" | "BCRT" => Some(Network::Regtest),
-            _ => None,
-        };
-        if let Some(network) = bech32_network {
-            let (_hrp, version, data) = bech32::segwit::decode(s)?;
-            let version = WitnessVersion::try_from(version).expect("we know this is in range 0-16");
-            let program = PushBytesBuf::try_from(data).expect("decode() guarantees valid length");
-            let witness_program = WitnessProgram::new(version, program)?;
+    fn from_str(s: &str) -> Result<Address<NetworkUnchecked>, ParseError> {
+        if let Ok((hrp, witness_version, data)) = bech32::segwit::decode(s) {
+            let version = WitnessVersion::try_from(witness_version)?;
+            let buf = PushBytesBuf::try_from(data).expect("bech32 guarantees valid program length");
+            let program = WitnessProgram::new(version, buf)
+                .expect("bech32 guarantees valid program length for witness");
 
-            return Ok(Address::new(network, Payload::WitnessProgram(witness_program)));
+            let hrp = KnownHrp::from_hrp(hrp)?;
+            let inner = AddressInner::Segwit { program, hrp };
+            return Ok(Address(inner, PhantomData));
         }
```

Comment:
Unrelated, for future PR, but I really don't like how the parsing code works. I think we should check prefix first and try to parse accordingly, then for unknown prefix emit some `UnknownFormatError`.

Reply:
Yes I agree, I think what I'm about to say is related, when we used the new bech32 the parsing code was more or less kept the same but the algo is now quite obviously improvable. I was a bit lazy and a bit wanting to keep scope limited to just using the bech32 lib.